File: examples/nirfmx/evdo/csharp/VS2010/RFmxEvdoChpSingleCarrier/vb/RFmxEvdoChpSingleCarrier.vb
'Steps:
'1. Open a new RFmx session.
'2. Configure Frequency Reference.
'3. Configure the basic signal properties (Center Frequency, Reference Level and External Attenuation).
'4. Configure Trigger Type and Trigger Parameters.
'5. Select CHP measurement and enable traces.
'6. Configure Sweep Time Parameters.
'7. Configure Averaging Parameters.
'8. Initiate the Measurement.
'9. Fetch CHP Measurements and Traces.
'10. Close the RFmx Session.

Imports NationalInstruments.RFmx.InstrMX
Imports NationalInstruments.RFmx.EvdoMX

Public Class RFmxEvdoChpSingleCarrier
    Private instrSession As RFmxInstrMX
    Private evdo As RFmxEvdoMX

    Private resourceName As String = "RFSA"

    Private measurement As RFmxEvdoMXMeasurementTypes = RFmxEvdoMXMeasurementTypes.Chp

    Private frequencyReferenceSource As String = RFmxInstrMXConstants.OnboardClock
    Private frequency As Double = 10000000.0                           ' Hz 

    Private centerFrequency As Double = 833490000.0                    ' Hz 
    Private externalAttenuation As Double = 0.0                        ' dB 
    Private digitalEdgeSource As String = RFmxEvdoMXConstants.Pfi0

    Private digitalEdge As RFmxEvdoMXDigitalEdgeTriggerEdge = RFmxEvdoMXDigitalEdgeTriggerEdge.Rising
    Private triggerDelay As Double = 0.0
    Private referenceLevel As Double = 0.0

    Private averagingEnabled As RFmxEvdoMXChpAveragingEnabled = RFmxEvdoMXChpAveragingEnabled.[False]
    Private averagingCount As Integer = 10
    Private averagingType As RFmxEvdoMXChpAveragingType = RFmxEvdoMXChpAveragingType.Rms

    Private sweepTimeAuto As RFmxEvdoMXChpSweepTimeAuto = RFmxEvdoMXChpSweepTimeAuto.[True]
    Private sweepTimeInterval As Double = 0.00167                    ' seconds 

    Private timeout As Double = 10.0                                 ' seconds 
    Private carrierAbsolutePower As Double                                  ' dBm 
    Private carrierRelativePower As Double                                  ' dB 
    Private spectrum As Spectrum(Of Single)

    Private enableAllTraces As Boolean = True
    Private enableTrigger As Boolean = False

    Public Sub Run()
        Try
            InitializeInstr()
            ConfigureEvdo()
            RetrieveResults()
            PrintResults()
        Catch ex As Exception
            DisplayError(ex)
        Finally
            CloseSession()
            Console.WriteLine("Press any key to exit")
            Console.ReadKey()
        End Try
    End Sub

    Private Sub InitializeInstr()
        ' Create a new RFmx Session 

        instrSession = New RFmxInstrMX(resourceName, "")
    End Sub

    Private Sub ConfigureEvdo()
        ' Get Evdo signal 

        evdo = instrSession.GetEvdoSignalConfiguration()

        ' Configure measurement 

        instrSession.ConfigureFrequencyReference("", frequencyReferenceSource, frequency)
        evdo.ConfigureRF("", centerFrequency, referenceLevel, externalAttenuation)
        evdo.ConfigureDigitalEdgeTrigger("", digitalEdgeSource, digitalEdge, triggerDelay, enableTrigger)
        evdo.SelectMeasurements("", measurement, enableAllTraces)
        evdo.Chp.Configuration.ConfigureSweepTime("", sweepTimeAuto, sweepTimeInterval)
        evdo.Chp.Configuration.ConfigureAveraging("", averagingEnabled, averagingCount, averagingType)
        evdo.Initiate("", "")
    End Sub

    Private Sub RetrieveResults()
        ' Retrieve results 

        evdo.Chp.Results.FetchCarrierMeasurement("", timeout, carrierAbsolutePower, carrierRelativePower)
        evdo.Chp.Results.FetchSpectrum("", timeout, spectrum)
    End Sub

    Private Sub PrintResults()
        Console.WriteLine("Carrier Absolute Power  (dBm) : {0}", carrierAbsolutePower)
    End Sub

    Private Sub CloseSession()
        If evdo IsNot Nothing Then
            evdo.Dispose()
            evdo = Nothing
        End If
        If instrSession IsNot Nothing Then
            instrSession.Close()
            instrSession = Nothing
        End If
    End Sub

    Private Shared Sub DisplayError(ex As Exception)
        Console.WriteLine("ERROR:" & vbLf & Convert.ToString(ex.[GetType]()) & ": " & ex.Message)
    End Sub

End Class

File: examples/nirfmx/evdo/csharp/VS2010/RFmxEvdoChpSingleCarrier/cs/RFmxEvdoChpSingleCarrier.cs
//Steps:
//1. Open a new RFmx session.
//2. Configure Frequency Reference.
//3. Configure the basic signal properties (Center Frequency, Reference Level and External Attenuation).
//4. Configure Trigger Type and Trigger Parameters.
//5. Select CHP measurement and enable traces.
//6. Configure Sweep Time Parameters.
//7. Configure Averaging Parameters.
//8. Initiate the Measurement.
//9. Fetch CHP Measurements and Traces.
//10. Close the RFmx Session.


using System;
using NationalInstruments.RFmx.InstrMX;
using NationalInstruments.RFmx.EvdoMX;

namespace NationalInstruments.Examples.RFmxEvdoChpSingleCarrier
{
    public class RFmxEvdoChpSingleCarrier
    {
        RFmxInstrMX instrSession;
        RFmxEvdoMX evdo;

        string resourceName = "RFSA";

        RFmxEvdoMXMeasurementTypes measurement = RFmxEvdoMXMeasurementTypes.Chp;

        string frequencyReferenceSource = RFmxInstrMXConstants.OnboardClock;
        double frequency = 10.0e+6;                     /* Hz */

        double centerFrequency = 833.49e+6;               /* Hz */
        double externalAttenuation = 0.00;          /* dB */
        string digitalEdgeSource = RFmxEvdoMXConstants.Pfi0;

        RFmxEvdoMXDigitalEdgeTriggerEdge digitalEdge = RFmxEvdoMXDigitalEdgeTriggerEdge.Rising;
        double triggerDelay = 0.00;
        double referenceLevel = 0.00;

        RFmxEvdoMXChpAveragingEnabled averagingEnabled = RFmxEvdoMXChpAveragingEnabled.False;
        int averagingCount = 10;
        RFmxEvdoMXChpAveragingType averagingType = RFmxEvdoMXChpAveragingType.Rms;

        RFmxEvdoMXChpSweepTimeAuto sweepTimeAuto = RFmxEvdoMXChpSweepTimeAuto.True;
        double sweepTimeInterval = 1.67e-3;             /* seconds */

        double timeout = 10.00;                     /* seconds */
        double carrierAbsolutePower;                /* dBm */
        double carrierRelativePower;                /* dB */
        Spectrum<float> spectrum;

        bool enableAllTraces = true;
        bool enableTrigger = false;

        public void Run()
        {
            try
            {
                InitializeInstr();
                ConfigureEvdo();
                RetrieveResults();
                PrintResults();
            }
            catch (Exception ex)
            {
                DisplayError(ex);
            }
            finally
            {
                CloseSession();
                Console.WriteLine("Press any key to exit");
                Console.ReadKey();
            }
        }

        private void InitializeInstr()
        {
            /* Create a new RFmx Session */
            instrSession = new RFmxInstrMX(resourceName, "");
        }

        private void ConfigureEvdo()
        {
            /* Get Evdo signal */

            evdo = instrSession.GetEvdoSignalConfiguration();

            /* Configure measurement */

            instrSession.ConfigureFrequencyReference("", frequencyReferenceSource, frequency);
            evdo.ConfigureRF("", centerFrequency, referenceLevel, externalAttenuation);
            evdo.ConfigureDigitalEdgeTrigger("", digitalEdgeSource, digitalEdge, triggerDelay, enableTrigger);
            evdo.SelectMeasurements("", measurement, enableAllTraces);
            evdo.Chp.Configuration.ConfigureSweepTime("", sweepTimeAuto, sweepTimeInterval);
            evdo.Chp.Configuration.ConfigureAveraging("", averagingEnabled, averagingCount, averagingType);
            evdo.Initiate("", "");
        }

        private void RetrieveResults()
        {
            /* Retrieve results */

            evdo.Chp.Results.FetchCarrierMeasurement("", timeout, out carrierAbsolutePower,
                out carrierRelativePower);
            evdo.Chp.Results.FetchSpectrum("", timeout, ref spectrum);
        }

        private void PrintResults()
        {
            Console.WriteLine("Carrier Absolute Power  (dBm) : {0}", carrierAbsolutePower);
        }

        void CloseSession()
        {
            if (evdo != null)
            {
                evdo.Dispose();
                evdo = null;
            }
            if (instrSession != null)
            {
                instrSession.Close();
                instrSession = null;
            }
        }

        static private void DisplayError(Exception ex)
        {
            Console.WriteLine("ERROR:\n" + ex.GetType() + ": " + ex.Message);
        }

    }
}
